Add BORT

stefan-it opened the pull request:
Hi,

this PR adds the recently introduced BORT model from @adewynter and Daniel J. Perry from the Alexa team into Transformers.

----

BORT was introduced in the [Optimal Subarchitecture Extraction For BERT](https://arxiv.org/abs/2010.10499).

Details about BORT:

> We extract an optimal subset of architectural parameters for the BERT architecture from Devlin et al. (2018) by applying recent breakthroughs in algorithms for neural architecture search. This optimal subset, which we refer to as "Bort", is demonstrably smaller, having an effective (that is, not counting the embedding layer) size of 5.5% the original BERT-large architecture, and 16% of the net size. Bort is also able to be pretrained in 288 GPU hours, which is 1.2% of the time required to pretrain the highest-performing BERT parametric architectural variant, RoBERTa-large (Liu et al., 2019), and about 33% of that of the world-record, in GPU hours, required to train BERT-large on the same hardware. It is also 7.9x faster on a CPU, as well as being better performing than other compressed variants of the architecture, and some of the non-compressed variants: it obtains performance improvements of between 0.3% and 31%, absolute, with respect to BERT-large, on multiple public natural language understanding (NLU) benchmarks.

This should fix #8135 :hugs: 

---

ToDo tasks:

* [x] Upload models (both PyTorch and TensorFlow model) to model hub
* [x] Add conversion script from Gluonnlp to Transformers
* [x] Enable unit tests (they are working and just wait for the model upload)

patrickvonplaten commented on tests/test_modeling_tf_bort.py:
```diff
@@ -0,0 +1,52 @@
+        input_ids = tf.convert_to_tensor(
+            [[0, 18077, 4082, 7804, 8606, 6195, 2457, 3321, 11, 10489, 16, 269, 2579, 328, 2]],
+            dtype=tf.int32,
+        )  # Schloß Nymphenburg in Munich is really nice!
```
Das stimmt!

patrickvonplaten commented on docs/source/model_doc/bort.rst:
```diff
@@ -0,0 +1,146 @@
+.. 
```
Doc looks great!

patrickvonplaten commented on src/transformers/models/bort/modeling_bort.py:
```diff
@@ -0,0 +1,143 @@
+# coding=utf-8
```
(nit) we could add some examples here as well similar to how it's done for MT5: https://github.com/huggingface/transformers/blob/c19d04623eacfbc2c452397a5eda0fde42db3fc5/src/transformers/models/mt5/modeling_mt5.py#L33

I think for `BortModel` we can just show how to get the `last_hidden_state` and for all other models we could show how to get the loss for fine-tuning.

patrickvonplaten reviewed (commented):
PR looks great! Think we only have to wait now for the name and then we're good to go :-)

julien-c commented:
🔥 Looking forward to taking a look at the conversion script from GluonNLP/mxnet!

stefan-it commented:
@patrickvonplaten I added some examples for both `modeling_bort.py` and modeling_tf_bort.py` :hugs: 

@julien-c The conversion script is also added - you just need to install  `gluonnlp==0.8.3` and `mxnet==1.5.0`.

These versions are defined in the BORT [requirements file](https://github.com/alexa/bort/blob/master/requirements.txt). The conversion script also performs a version check.

patrickvonplaten commented on src/transformers/models/bort/convert_bort_original_gluonnlp_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,316 @@
+        gluon_param = to_torch(params[gluon_param])
+        shape_gluon = gluon_param.shape
+
+        assert shape_hf == shape_gluon
```
maybe a small assert message would be great just:

```suggestion
        assert shape_hf == shape_gluon, f"The gluon parameter {gluon_param} has shape {shape_gluon}, but expects shape {shape_hf} for Transformers"
```

patrickvonplaten commented on src/transformers/models/bort/modeling_bort.py:
```diff
@@ -0,0 +1,231 @@
+        >>> hidden_states = outputs.last_hidden_state
+    """
+
+    config_class = BortConfig
```
we should add `model_type = 'bort'` for each class here -> see MT5 for comparison: https://github.com/huggingface/transformers/blob/51adb97cd644a5840d971868d18c1d436fd6ff5d/src/transformers/models/mt5/modeling_mt5.py#L43

patrickvonplaten commented on src/transformers/models/bort/modeling_tf_bort.py:
```diff
@@ -0,0 +1,250 @@
+    """
+
+    config_class = BortConfig
+
```
also add model type here for all models: https://github.com/huggingface/transformers/blob/51adb97cd644a5840d971868d18c1d436fd6ff5d/src/transformers/models/mt5/modeling_tf_mt5.py#L65

patrickvonplaten reviewed (commented):
Really cool! Have one (nit) and I think we should add the model_type to each class

patrickvonplaten reviewed (approved):
Looks great!

patrickvonplaten commented:
We'll have to think a bit how to advertise this. Let me draft up a "Contribution Proposal" for the fine-tuning algorithm.

sgugger reviewed (approved):
Thanks for adding this model! There are a few things to adapt to have the same API as the current master and I would very much like to be consistent with the paper and use Bort (not BORT) everywhere in the docs.

LysandreJik commented on docs/source/model_doc/bort.rst:
```diff
@@ -0,0 +1,146 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. autoclass:: transformers.BortTokenizerFast
+    :members:
```
I think @sgugger made a typo here, I think you can leave it as `:members:` given that it's a fast tokenizer.

sgugger replied:
Sorry! I meant it for the PyTorch models :-)

LysandreJik reviewed (commented):
Hi @stefan-it, thanks a lot for your contribution!

If Bort can be loaded seamlessly in the BERT architecture, is there really a need to redefine all models in PyTorch and TensorFlow? We would need to do this for all models on the hub if that was the case. If there was a change in one of the models I would understand, but given that it's an exact copy if BERT I don't think that's necessary at all.

I understand the conversion script, however. I would just replace the modelname "bort" by "bert" in that model script so that the models are loadable directly in the BERT architecture.

I see that Bort requires RoBERTa tokenizers, which isn't a problem either; tokenizers can be decoupled from their models by specifying a `tokenizer_class` in the model config, similarly to what [BERTweet does.](https://huggingface.co/vinai/bertweet-base/blob/main/config.json)

patrickvonplaten commented:
Hey @stefan-it,

I've discussed a bit with @LysandreJik and @sgugger offline and I do agree with @LysandreJik after having thought about it again. I think it's better if we actually don't add any new code (besides the conversion script) that should be added to `src/transformers/models/bert/` and the docs page. I'm very sorry to have you asked to go down this road! I think however it does make more sense to not add any "tokenizer" or "model" code as those are exact copies of the `RobertaTokenizer` and `BertModel`. It's probably most efficient to open a new PR and only add the required files. Super sorry again!

gaceladri commented:
Are we planning to implement the architectural optimization (FPTAS) or just the pre-trained models?

patrickvonplaten commented:
> Are we planning to implement the architectural optimization (FPTAS) or just the pre-trained models?

Great question! For now, we'll just add the model weights - see: #9813. A community contribution showing how to do FPTAS in a notebook would be extremely valuable though.

patrickvonplaten commented:
Closing in favor of #9813
